v2/writer.go

```diff
@@ -0,0 +1,138 @@
+package car
+
+import (
+	"bytes"
+	"context"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	carv1 "github.com/ipld/go-car"
+	"github.com/willscott/carbs"
+	"io"
+)
+
+const bulkPaddingBytesSize = 1024
+
+var bulkPadding = make([]byte, bulkPaddingBytesSize)
+
+type (
+	// padding represents the number of padding bytes.
+	padding uint64
+	// Writer writes CAR v2 into a give io.Writer.
+	Writer struct {
+		Walk         carv1.WalkFunc
+		IndexCodec   carbs.IndexCodec
+		NodeGetter   format.NodeGetter
+		CarV1Padding uint64
+		IndexPadding uint64
+
+		ctx          context.Context
+		roots        []cid.Cid
+		encodedCarV1 *bytes.Buffer
+	}
+)
+
+// WriteTo writes this padding to the given writer as default value bytes.
+func (p padding) WriteTo(w io.Writer) (n int64, err error) {
+	var reminder int64
+	if p > bulkPaddingBytesSize {
+		reminder = int64(p % bulkPaddingBytesSize)
+		iter := int(p / bulkPaddingBytesSize)
+		for i := 0; i < iter; i++ {
+			if _, err = w.Write(bulkPadding); err != nil {
+				return
+			}
+			n += bulkPaddingBytesSize
+		}
+	} else {
+		reminder = int64(p)
+	}
+
+	paddingBytes := make([]byte, reminder)
+	_, err = w.Write(paddingBytes)
+	n += reminder
+	return
+}
+
+// NewWriter instantiates a new CAR v2 writer.
+// The writer instantiated uses `carbs.IndexSorted` as the index codec,
+// and `carv1.DefaultWalkFunc` as the default walk function.
+func NewWriter(ctx context.Context, ng format.NodeGetter, roots []cid.Cid) *Writer {
+	return &Writer{
+		Walk:         carv1.DefaultWalkFunc,
+		IndexCodec:   carbs.IndexSorted,
+		NodeGetter:   ng,
+		ctx:          ctx,
+		roots:        roots,
+		encodedCarV1: new(bytes.Buffer),
```

Comment:
@masih I'm assuming that this is an initial, easy version and that buffering the entirety of CARv1 in memory before writing isn't something we're planning on being stuck with? One of the nice features of using fixed byte length blocks at the start (as opposed to the usual varint length blocks that we do most other places around our stack) is the ability to come back and update the header details _after_ writing CARv1 bytes.

Reply:
Absolutely; all the copied car content will go away in the coming PRs. I have left a bunch of TODOs in code; it depends on carbs API refactoring in places.